### PR TITLE
Ability to scale eigenvector in Eigenvalue executioner

### LINKED ISSUE #14500: Scale eigenvector with new Eigenvalue executioner
## Reason
Typically the solution to an eigenvalue problem, or eigenvector, can be scaled by anything and still be correct. I believe the SLEPc eigenvalue solver scales the solution to the L2 norm of the solution vector. This type of normalization does not have much physical meaning. However, when using eigenvalue solves to initialize transient simulations or coupling multi-physics, the scale of the eigenvector matters significantly. It would useful to have an input parameter in the Eigenvalue executioner that would allow the eigenvector to be scaled to a user defined quantity after the solve has finished.

## Design
After the executioner finishes the solve step, call a function in EigenProblem that scales the variables that have been marked as eigen variables. The scaling would include an inputted postprocessor value and a scaler, which is to have the ability to scale the eigenvector such that a certain postprocessor is a specified value.

## Impact
Allow users to transfer and couple an eigenvector with a physically meaningful normalization.

### PR DESCRIPTION
**Pull request by zachmprince**
This PR adds two new parameters to the Eigenvalue executioner: `normalization` and `normal_factor`. These parameters are effectively the same as those of the same name in `EigenExecutionerBase`. `normalization` defines an postprocessor to scale the eigenvector and `normal_factor` multiplies the eigenvector. In practice, the solution should will be scaled such that the resulting postprocessor values is equal to the inputed `normal_factor` value.

closes #14500

### CODE REVIEW

**Comment by zachmprince** [comment]
I'm not sure why these tests are failing. I can't seem to reproduce them.

**Comment by zachmprince** [comment]
@fdkong could you help me figure out why these tests are failing?

**Comment by fdkong** [comment]
@zachmprince At this point, I think you should just safeguard your code.

In the future, we need to enhance libMesh eigensystem.

**Comment by zachmprince** [comment]
> @zachmprince At this point, I think you should just safeguard your code.

Thank you @fdkong. Did a quick patch, hopefully it passes now.

**Comment by zachmprince** [comment]
@fdkong looks like we are good to go. Is my implementation all good?

File: framework/src/executioners/Eigenvalue.C [hunk 1]
```diff
@@ -71,3 +76,24 @@ Eigenvalue::execute()
 
   Steady::execute();
 }
+
+void
+Eigenvalue::postSolve()
+{
+  Real val = getParam<Real>("normal_factor");
```
Comment (YaqiWang) [inline framework/src/executioners/Eigenvalue.C]: Looks to me users can set normal_factor without giving the postprocessor name. In such a case, the value will be used for scaling directly. This line needs to be moved into the following if branch. A separate scaling factor is declared here initialized to 1.

**Review by YaqiWang — CHANGES_REQUESTED** [review]

File: framework/src/executioners/Eigenvalue.C [hunk 2]
```diff
@@ -28,6 +28,9 @@ Eigenvalue::validParams()
 
   params.addPrivateParam<bool>("_use_eigen_value", true);
 
+  params.addParam<PostprocessorName>("normalization", "To evaluate |x| for normalization");
```
Comment (fdkong) [inline framework/src/executioners/Eigenvalue.C]: postprocessor  to evaluate .....

File: framework/src/executioners/Eigenvalue.C [hunk 3]
```diff
@@ -28,6 +28,9 @@ Eigenvalue::validParams()
...
+  params.addParam<Real>("normal_factor", 1.0, "Normalize x to make |x| equal to this factor");
+
```
Comment (fdkong) [inline framework/src/executioners/Eigenvalue.C]: We possibly should replace x with eigenvector?

File: framework/src/executioners/Eigenvalue.C [hunk 4]
```diff
@@ -49,6 +54,10 @@ Eigenvalue::Eigenvalue(const InputParameters & parameters)
   Moose::SlepcSupport::storeSlepcEigenProblemOptions(_eigen_problem, parameters);
   _eigen_problem.setEigenproblemType(_eigen_problem.solverParams()._eigen_problem_type);
 #endif
+
+  if (!parameters.isParamValid("normalization") && parameters.isParamSetByUser("normal_factor"))
+    paramError("normal_factor",
+               "Cannot set scaling factor without defining normalization postprocessor.");
```
Comment (fdkong) [inline framework/src/executioners/Eigenvalue.C]: We actually should allow users to set an arbitrary scaling number if they want.

File: framework/src/problems/EigenProblem.C [hunk 5]
```diff
@@ -294,3 +294,27 @@ EigenProblem::isNonlinearEigenvalueSolver()
...
          solverParams()._eigen_solve_type == Moose::EST_MF_MONOLITH_NEWTON;
 }
+
+void
+EigenProblem::scaleEigenvector(const Real scaling_factor)
+{
+// FIXME: This needs to be protected since NonlinearEigenSystem derives from ParallelObject (instead
+// of NonlinearSystemBase) when SLEPc isn't installed
+#if LIBMESH_HAVE_SLEPC
+  std::vector<VariableName> var_names = getVariableNames();
```
Comment (fdkong) [inline framework/src/problems/EigenProblem.C]: Are you scaling all variables? or just eigen variables? From my understanding, we should scale only Eigen variables

**Review by fdkong — CHANGES_REQUESTED** [review]

File: framework/src/problems/EigenProblem.C (hunk 5 above)
Comment (zachmprince) [reply framework/src/problems/EigenProblem.C]: Only variables marked eigen. You can see the if statement in the following line.

File: framework/src/executioners/Eigenvalue.C (hunk 4 above)
Comment (zachmprince) [reply framework/src/executioners/Eigenvalue.C]: The scaling factor doesn't make much sense if it isn't related to some physical quantity provided by the postprocessor. So @YaqiWang and I think that preventing the user from only scaling the solution with this factor is important. That being said, a user could theoretically trick it by just inputting a constant postprocessor.

**Review by fdkong — APPROVED** [review]

**Comment by fdkong** [comment]
It looks great

**Comment by permcody** [comment]
@rwcarlsen - I'm seeing a Pronghorn diff on this. I don't think it's real, can you confirm?

**Comment by rwcarlsen** [comment]
It doesn't look like a particularly significant failure, but Pronghorn doesn't use any EigenExecutioner stuff and Pronghorn doesn't seem to be failing on next - so there is something funky going on.

**Comment by fdkong** [comment]
> Job [App tests PETSc alt](https://civet.inl.gov/job/452533/) on [45a319a](https://github.com/idaholab/moose/commit/45a319aed1942b8b86c02e23f1d3872b3687bd3a) : invalidated by @fdkong
> 
> Mammoth? tutorials/basic_mammoth_components/meshing.exercise_ms_4

Obviously, this kind of falling is random

**Comment by YaqiWang** [comment]
Nah, it is something we have fixed due to anther PR. It is not random, ;-)

**Pull request merged**